server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +345,266 @@ public DirectoryReader getDirectoryReader() {
+            if (pointValues == null) {
+                return null;
+            }
+            return (cancellable.get() != null) ? new ExitablePointValues(pointValues, cancellable.get()) : pointValues;
+        }
+
+        @Override
+        public Terms terms(String field) throws IOException {
+            Terms terms = in.terms(field);
+            if (terms == null) {
+                return null;
+            }
+            return (cancellable.get() != null && terms instanceof CompletionTerms == false) ?
+                    new ExitableTerms(terms, cancellable.get()) : terms;
+        }
+
+        @Override
+        public CacheHelper getCoreCacheHelper() {
+            return in.getCoreCacheHelper();
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            return in.getReaderCacheHelper();
+        }
+    }
+
+    /**
+     * Helper class to be used as an immutable reference so that the underlying
+     * {@link QueryCancellable} passed trough the hierarchy to the {@link Terms} and {@link PointValues}
+     * during construction can be set later with {@link ContextIndexSearcher#setCancellable}
+     */
+    private static class Holder<T> {
+
+        private T in;
+
+        private Holder(T in) {
+            this.in = in;
+        }
+
+        private void set(T in) {
+            this.in = in;
+        }
+
+        private T get() {
+            return in;
+        }
+    }
+
+    /**
+     * Wrapper class for {@link FilterTerms} that check for query cancellation or timeout.
+     */
+    static class ExitableTerms extends FilterTerms {
+
+        private final QueryCancellable cancellable;
+
+        private ExitableTerms(Terms terms, QueryCancellable cancellable) {
+            super(terms);
+            this.cancellable = cancellable;
+        }
+
+        @Override
+        public TermsEnum intersect(CompiledAutomaton compiled, BytesRef startTerm) throws IOException {
+            return new ExitableTermsEnum(in.intersect(compiled, startTerm), cancellable);
+        }
+
+        @Override
+        public TermsEnum iterator() throws IOException {
+            return new ExitableTermsEnum(in.iterator(), cancellable);
+        }
+    }
+
+    /**
+     * Wrapper class for {@link FilterTermsEnum} that is used by {@link ExitableTerms} for
+     * implementing an exitable enumeration of terms.
+     */
+    private static class ExitableTermsEnum extends FilterTermsEnum {
+
+        private final QueryCancellable cancellable;
+
+        private ExitableTermsEnum(TermsEnum termsEnum, QueryCancellable cancellable) {
+            super(termsEnum);
+            this.cancellable = cancellable;
+            this.cancellable.checkCancelled();
+        }
+
+        @Override
+        public BytesRef next() throws IOException {
+            // Before every iteration, check if the iteration should exit
+            this.cancellable.checkCancelled();
```

Comment:
Can we do it on a regular interval (like in ExitableIntersectVisitor) instead of for every term, I worry that this might add noticeable overhead.